src/types/sign_self.rs

```diff
@@ -0,0 +1,25 @@
+use super::*;
+use crate::{crypto::verify_proof, error::Result};
+
+pub trait SignSelf {
+    type Extra;
+    fn msg_hash(&self, chain_id: &str, contract_addr: &str, extra: Self::Extra) -> Result<Hash>;
+
+    fn sign(&self, signing_key: &[u8], msg_hash: &[u8]) -> Result<Vec<u8>> {
+        let proof = VRF.prove(signing_key, msg_hash)?;
+
+        Ok(proof)
+    }
+
+    fn verify(
+        &self,
+        public_key: &[u8],
+        proof: &[u8],
```

Comment:
I'm probably missing something but it seems that for all messages we first get the public key and proof from `self` and then pass them to `self.verify()`.

Example:
```rs
let chain_id = CHAIN_ID.load(deps.storage)?;
        let proof = Vec::<u8>::from_hex_str(&self.proof)?;
        let public_key = PublicKey::from_hex_str(&self.public_key)?;
        self.verify(
            &public_key,
            &proof,
            &chain_id,
            env.contract.address.as_str(),
            inc_get_seq(deps.storage, &public_key)?,
        )?;
```

Can't the `verify` function take care of the public key and proof similar to the hash?